### PR TITLE
Gal 643/mesh module

### PR DESCRIPTION
**Pull request by msandhuNHS**
## Description

Added service layer. A mesh service can be imported via a npm module to communicate with mesh-mailbox.

1. Added service classes
2. Added unit test

## Type of changes

- [ ] Refactoring (non-breaking change)
- [y] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: run.js
Comment (chenglawson1) [inline run.js:1]: Is this file needed by the module? Looks like test code?

File: src/model/payload.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default class payload {
```
Comment (chenglawson1) [inline src/model/payload.js:1]: Name convention is class names should be camel case, each word should start with an uppercase letter. This comment applies to all classes in this PR.

I appreciate this may be inherited from Andrew's prototype.

File: src/service/loader.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+export default class setup {
+  constructor(dotenv) {
+    let logLevel = process.env.LOG_LEVEL || "DEBUG";
+    log.setLevel(log.levels[logLevel]);
+
+    const result = dotenv.config();
+    if (result.error) {
+      throw result.error;
+    }
+    console.log("Environment variables loaded.");
```
Comment (chenglawson1) [inline src/service/loader.js:14]: Why is there a mixture of loglevel and console?

File: src/service/loader.js
Comment (zainm97) [inline src/service/loader.js:1]: everything coming from a process.env should be declared outside the class and read in as a constant

File: src/service/loader.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        key: readFileSync(process.env.MESH_SENDER_KEY_LOCATION),
+        rejectUnauthorized: false,
+      });
+      this.receiverAgent = new Agent({
+        cert: readFileSync(process.env.MESH_RECEIVER_CERT_LOCATION),
+        key: readFileSync(process.env.MESH_RECEIVER_KEY_LOCATION),
+        rejectUnauthorized: false,
+      });
+    };
+    this.senderMailboxID = process.env.MESH_SENDER_MAILBOX_ID || "X26ABC1";
```
Comment (zainm97) [inline src/service/loader.js:41]: X26ABC1 is this meaningful in any way? If yes we shouldn't be hardcoding. If no, would be more helpful to shoud that its a placeholders i.e "A00AAA0"

File: src/service/loader.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    const result = dotenv.config();
+    if (result.error) {
+      throw result.error;
+    }
+    console.log("Environment variables loaded.");
+
+    this.url = process.env.MESH_URL || "https://localhost:8700";
+    this.sharedKey = process.env.MESH_SHARED_KEY || "TestKey";
+    this.sandbox = process.env.MESH_SANDBOX || "true";
```
Comment (zainm97) [inline src/service/loader.js:18]: true should be a bool not a string

File: src/service/helper.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+// A 30 second wait timer to allow the messages to be received and processed by mesh
+// Override this value by passing custom seconds when invoked
+export async function waitThirtySeconds(override) {
+  log.debug("\nChecking connection to mailbox with handshake");
+  return new Promise((resolve) => {
+    if (override === undefined) {
+      setTimeout(() => {
+        resolve("30 seconds have passed.");
+      }, 30000); // 30000 milliseconds = 30 seconds
```
Comment (zainm97) [inline src/service/helper.js:15]: dont need clarifcation for ms. In docs of MDN and your method has specifed seconds so this conversion is assumed by caller of function.

File: src/service/meshService.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  async sendFile() {
+    log.debug("\nsending a file");
+    await this.senderService.createAndSendMessageChunks();
+  }
+
+  async receiveMessage(wait) {
+    if(wait === true){
+      log.debug("\nwaiting 30 seconds for mesh to process the message");
+      await waitThirtySeconds();
```
Comment (zainm97) [inline src/service/meshService.js:29]: you have given waitThirtySeconds the ability to override in its definition. Its only called once but doesn have the ability to override as you havent supplied params in receiveMessage that can be passed down to waitThirtySeconds

File: src/service/receiverService.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+          mailboxPassword: this.mailboxPassword,
+          sharedKey: this.sharedKey,
+          messageID: message,
+          agent: this.agent
+        })
+        try {
+          if (messageResponse.data === "") {
+            log.warn("WARNING: No data for message " + message);
+          } else if (messageResponse.status === 200) {
+            log.debug("Message ID is: " + message);
```
Comment (zainm97) [inline src/service/receiverService.js:60]: the logic in this condition statement is the same as the one below. Either create a function to move out duplicate code or adjust the conditional statement. When we incorporate sonarqube into this it will flag it up as duplicate code

File: tests/meshService.test.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+import receiverService from "../src/service/receiverService.js"
+import meshService from "../src/service/meshService.js";
+
+// Empty input and output dirs
+async function emptyDirs() {
+  const inputDir = "./input";
+  const outputDir = "./output";
+
+  fs.readdir(inputDir, (err, files) => {
+    if (err) throw err;
```
Comment (zainm97) [inline tests/meshService.test.js:17]: duplication of code with below function. Best to create a function that takes directory as a argument. Sonarqube will complain otherwise

File: src/service/helper.js (hunk 5 above)
Comment (chenglawson1) [reply src/service/helper.js:15]: To be fair this might be inherited from Andrew's prototype.

File: src/service/loader.js (hunk 4 above)
Comment (chenglawson1) [reply src/service/loader.js:18]: Environment variables are strings by default, hence the default is a string "true".

File: src/service/loader.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    if (this.sandbox === true) {
```
[the remark this answers is not in the document]
Comment (chenglawson1) [reply src/service/loader.js:19]: this.sandbox looks like a string variable, shouldn't it be === "true"?

[the remark this answers is not in the document]
Comment (zainm97) [reply src/service/loader.js:19]: cool, as commented below happy to resolve comment

File: run.js
Comment (msandhuNHS) [reply run.js:1]: I should be able to exclude it when publishing, will check

File: src/model/payload.js (hunk 1 above)
Comment (msandhuNHS) [reply src/model/payload.js:1]: Sure, I was following his conventions, do you prefer we change?

File: src/service/helper.js (hunk 5 above)
Comment (msandhuNHS) [reply src/service/helper.js:15]: Can make the timeout an environmental variable

File: src/service/loader.js
Comment (zainm97) [reply src/service/loader.js:1]: its a constant. Make it available globally

File: src/service/loader.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (msandhuNHS) [reply src/service/loader.js:19]: Have corrected, thanks for spotting

File: src/service/loader.js (hunk 3 above)
Comment (msandhuNHS) [reply src/service/loader.js:41]: These are ids Andrew is using

File: src/service/meshService.js (hunk 6 above)
Comment (msandhuNHS) [reply src/service/meshService.js:29]: If a value is provided then it will override the default 30 seconds. Andrew tested and found 30 seconds to be ideal for sending and receiving a file. This may be useful in the future incase we need a longer timeout.

File: src/service/receiverService.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                if (err) {
+                  log.error(
+                    `ERROR: an error occurred while trying to write chunk data: ${err}`
+                  );
+                }
+              }
+            );
+          }
+        } catch {
+          console.error("ERROR: Failure reading message" + message);
```
[the remark this answers is not in the document]
Comment (msandhuNHS) [reply src/service/receiverService.js:91]: This is Andrew's implementation

File: src/service/receiverService.js (hunk 7 above)
Comment (msandhuNHS) [reply src/service/receiverService.js:60]: Andrew's implementation, will speak with him and let him know

File: tests/meshService.test.js (hunk 8 above)
Comment (msandhuNHS) [reply tests/meshService.test.js:17]: Ok will do

Comment (msandhuNHS) [reply tests/meshService.test.js:17]: Is this considered duplicate code by sonarqube?

File: src/service/receiverService.js (hunk 7 above)
Comment (chenglawson1) [reply src/service/receiverService.js:60]: Andrew's code was only a prototype, when you modularised it it should be tidied up.

File: src/service/loader.js (hunk 3 above)
Comment (msandhuNHS) [reply src/service/loader.js:42]: the hardcoded Ids run when sandbox is toggled

File: src/service/meshService.js (hunk 6 above)
Comment (msandhuNHS) [reply src/service/meshService.js]: timeout is now an env variable

File: run.js
Comment (chenglawson1) [inline run.js:1]: I thought I asked if this is test code that is not required in the module?

File: tests/meshService.test.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+describe('mesh service', () => {
+
+  // Load variables
+  const loaderInstance = new loader(dotenv);
+
+  const messageFile =
+    process.env.MESH_DATA_FILE || "./tests/testdata-organizations-100000.csv";
+
+  // Create payload
+  const data = new payload(
```
Comment (chenglawson1) [inline tests/meshService.test.js:35]: Class is now Payload not payload, please check other classes you've renamed as well and retest before confirming PR is ready for review again.

File: src/service/loader.js (hunk 3 above)
Comment (chenglawson1) [reply src/service/loader.js:42]: Zain is concerned that mail box id being exposed in a public repo is not a good idea, and I agree.

I think maybe the best approach is if a mail box id is not setup in the environment variable an error should be thrown/returned rather than using a hardcoded sandbox mail box id.

File: tests/meshService.test.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    "This is a JEST test.",
+    messageFile
+  )
+
+  // Get destination Id or provide one
+  const destination = loaderInstance.receiverMailboxID || "X26OT264";
+
+  // Configure sender service
+  const sendInstance = new senderService(
```
Comment (chenglawson1) [inline tests/meshService.test.js:44]: Class name renamed.

File: tests/meshService.test.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    loaderInstance.senderConfig(),
+    data,
+    destination);
+
+  // Configure receiver service
+  const receiverInstance = new receiverService(
```
Comment (chenglawson1) [inline tests/meshService.test.js:50]: Class name issue.

File: src/service/receiverService.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (chenglawson1) [reply src/service/receiverService.js:106]: Good that you have introduced switch statement, but should have a default that throws an error to be caught by catch block.
Unless you are certain those are the only http status code that can be returned?

File: run.js
Comment (msandhuNHS) [reply run.js:1]: Will remove this when publishing

File: src/service/loader.js (hunk 3 above)
Comment (msandhuNHS) [reply src/service/loader.js:42]: They are already visible in index.js within the repo, I can remove them in my files, would you also want them removed from the entire repo?

File: src/service/receiverService.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (msandhuNHS) [reply src/service/receiverService.js:106]: I can introduce it, I'm not certain

File: src/service/loader.js (hunk 3 above)
Comment (msandhuNHS) [reply src/service/loader.js:42]: The ID’s are for the defaults that are created when you generate a local sandbox, those values were gathered from another NHS git repo so anyone that wants to know them can just go a look at the mesh-sandbox github repo

File: src/service/senderService.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
+import log from "loglevel";
+import sendMessage from "../post/send_message.js";
+import sendMessageChunks from "../post/send_message_chunks.js";
+import { checkConnection } from "./Helper.js";
+import payload from "../model/Payload.js";
```
Comment (chenglawson1) [inline src/service/senderService.js:5]: Importing lowercase payload does not look right, since class is not Payload. Don't think it is being used anyway.

**Review by chenglawson1 — APPROVED** [review]

**Review by andrew-cleveland — APPROVED** [review]
